### PR TITLE
Add design documentation for profiling tool.

### PR DESCRIPTION
**Pull request by qingqing01**
Fix https://github.com/PaddlePaddle/Paddle/issues/6699

For a better review, you could see it [here](https://github.com/qingqing01/Paddle/blob/profiler_design/doc/design/profiler.md).

### CODE REVIEW

File: doc/design/profiler.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+### Event
+
+In above work flow, a pair of events are needed before and aftern the piece of code to collect time. So the event has a flag to mark it is starting event or ending event. There three kinds of event:
+
+```c++
+enum EventKind { kMark,
+  kPushRange,
+  kPopRange};
+```
+- kMark: only a mark.
```
Comment (reyoung) [inline doc/design/profiler.md]: I do not understand `what is only a mark`.

File: doc/design/profiler.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+The overall flow is shown as following figure.
+
+<img src="./images/profiler.png" align="center"/><br/>
+
+### Event
+
+In above work flow, a pair of events are needed before and aftern the piece of code to collect time. So the event has a flag to mark it is starting event or ending event. There three kinds of event:
+
+```c++
+enum EventKind { kMark,
```
Comment (reyoung) [inline doc/design/profiler.md]: Better Indent

File: doc/design/profiler.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+As mentioned above, there is no need to record the timeline when disabling the profiler. So there is a global state to enable or distable the profiler. 
+
+```c++
+enum ProfilerState {
+  kDisabled, 
+  kCPU,
+  kCUDA
+};
+ProfilerState kState;
```
Comment (reyoung) [inline doc/design/profiler.md]: g_state;

File: doc/design/profiler.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+## Architecture
+
+The work flow for most task is as follows. Each operator will run many times in the all iterations. So the profiler must collect the total time of each operator during the iteration. For more, sometimes, the developers want to collect more detailed time span inside the operator or record time span for elsewhere, this requires that the profiler must support to record the nested time span. And in order to speed training, all the deep learning framework supports parallel computing, including multi-threads on CPU and multi-GPUs. So the profiler must enable to collect the timeline for each thread. In addition, the profiler also occupies certain resources. It must can be easily to enable or disable by the developers. At last, the profiler should show a human-readable report.
```
Comment (kuke) [inline doc/design/profiler.md]: ```speed``` -> ```speedup```
```multi-threads```, ```multi-GPUs``` -> ```multiple threads```, ```multiple GPUs```. They should be nouns here.
```enable``` -> ```be able```
```It must can be easily to enable or disable by the developers``` -> ```It must be easily to be enabled or disabled by the developers```
```show``` -> ```present```

File: doc/design/profiler.md [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+## Architecture
+
+The work flow for most task is as follows. Each operator will run many times in the all iterations. So the profiler must collect the total time of each operator during the iteration. For more, sometimes, the developers want to collect more detailed time span inside the operator or record time span for elsewhere, this requires that the profiler must support to record the nested time span. And in order to speed training, all the deep learning framework supports parallel computing, including multi-threads on CPU and multi-GPUs. So the profiler must enable to collect the timeline for each thread. In addition, the profiler also occupies certain resources. It must can be easily to enable or disable by the developers. At last, the profiler should show a human-readable report.  
+
+```python
+for i in xrange(M):  # M is  the iteration number
+  for op in operator_lists: # The `operator_lists` is the all operators in the network graph.
```
Comment (kuke) [inline doc/design/profiler.md]: ```is the all operators``` -> ```contains all the operators```

File: doc/design/profiler.md [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    op.run();
+```
+
+In a summary, the proflier should have follow features:
```
Comment (kuke) [inline doc/design/profiler.md]: ```In a summary``` -> ```In summary```
```follow features``` -> ```the following features```

File: doc/design/profiler.md [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+- record time span in loop.
```
Comment (kuke) [inline doc/design/profiler.md]: ```record``` -> ```records``` 
```support``` -> ```supports```

File: doc/design/profiler.md [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+- support nested time span.
+- support multi-threads/multi-GPUs.
+- support to enable and disable the profiler.
```
Comment (kuke) [inline doc/design/profiler.md]: ```support to enable and disable the profiler.``` -> ```supports to be enabled and disabled.```

File: doc/design/profiler.md [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+But how to record the time for the mixed C++ and CUDA program?  There many C++ interfaces to get the current calendar time in host program. But for GPU, the CUDA kernels may be executed concurrently if they are in different streams (http://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#streams) and the CUDA kernels is asynchronous with the host program if there is no the synchronous aftern the CUDA kernels. The CUDA provides [event](http://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#events) to monitor the device's perform accurate timing. Inspired by PyTorch and CUDA event, we also design and apply the events to record the timeline.  Then summary and show statistics based on these events.
```
Comment (kuke) [inline doc/design/profiler.md]: ```interfaces``` -> ```APIs```?
```there is no the synchronous aftern the CUDA kernels``` -> ```there is no synchronization  between different CUDA kernels```
```The CUDA``` -> ```CUDA```
```monitor the device's perform accurate timing``` -> ```monitor the device and perform accurate timing```
```summary``` -> ```summarize```

File: doc/design/profiler.md [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+In a summary, the proflier should have follow features:
+
+- record time span in loop.
+- support nested time span.
+- support multi-threads/multi-GPUs.
+- support to enable and disable the profiler.
+
+But how to record the time for the mixed C++ and CUDA program?  There many C++ interfaces to get the current calendar time in host program. But for GPU, the CUDA kernels may be executed concurrently if they are in different streams (http://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#streams) and the CUDA kernels is asynchronous with the host program if there is no the synchronous aftern the CUDA kernels. The CUDA provides [event](http://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#events) to monitor the device's perform accurate timing. Inspired by PyTorch and CUDA event, we also design and apply the events to record the timeline.  Then summary and show statistics based on these events.  
+
+The overall flow is shown as following figure.
```
Comment (kuke) [inline doc/design/profiler.md]: ```as following figure``` -> ```as the following figure```

File: doc/design/profiler.md [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+
+<img src="./images/profiler.png" align="center"/><br/>
+
+### Event
+
+In above work flow, a pair of events are needed before and aftern the piece of code to collect time. So the event has a flag to mark it is starting event or ending event. There three kinds of event:
```
Comment (kuke) [inline doc/design/profiler.md]: ```aftern``` -> ```after```
```to collect time``` -> ```to be analyzed```
```mark it is starting event or ending event``` -> ```mark whether it is a starting event or an ending event```
```There ``` -> ```There are```

File: doc/design/profiler.md [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+
+```c++
+enum EventKind { kMark,
+  kPushRange,
+  kPopRange};
+```
+- kMark: only a mark.
+- kPushRange: mark the starting event for time range. 
+- kPopRange: mark the ending event for the time range.
```
Comment (kuke) [inline doc/design/profiler.md]: ```the time range``` -> ```time range``` for consistency

File: doc/design/profiler.md [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+
+For the CPU code, the events only need to record the current time. For the CUDA code, the [event management functions of CUDA](http://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EVENT.html#group__CUDART__EVENT) are used.  For many pieces of code, a event lists are used to record each piece.
```
Comment (kuke) [inline doc/design/profiler.md]: ```a event lists are used to record each piece```-> ```an event list is used to record their timeline```

File: doc/design/profiler.md [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+  int device_ = -1;
+#endif
+};
+
+struct EventList {
+  std::forward_list<std::vector<Event>> event_blocks;
+};
+```
+
+As mentioned above, there is no need to record the timeline when disabling the profiler. So there is a global state to enable or distable the profiler.
```
Comment (kuke) [inline doc/design/profiler.md]: ```distable``` -> ```disable```

**Review by kuke — COMMENTED** [review]
A great design doc. Thanks!

File: doc/design/profiler.md [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
...p learning framework, the PaddlePaddle also used C++, CUDA and Python as the basic programming languages to adapt to run on CPU and GPU device.  The [`nvprof` tools](http://docs.nvidia.com/cuda/profiler-users-guide/index.html#nvprof-overview) is usually used to analyse the CUDA program.  We have [a document](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/howto/optimization/cpu_profiling.md) to profile CPU and Python program by [yep](https://pypi.python.org/pypi/yep) and [Google's perftools](https://github.com/google/pprof) to profile the only CPU and Python program. But for [PaddlePaddle fluid](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/fluid.md), the operator is the basic computing unit. The developers usually wants to collect the time of each operator and locate bottlenecks.  The `nvprof` usually collect the timeline of CUDA-related activities on both CPU and GPU, including kernel execution, memory transfers, memory set and CUDA API calls and events or metrics for CUDA kernels. And the `yep` and `Google's perftools` can't collect the timeline for CUDA program. All these tools can't collect time in the operator level. So we design this profiling tools.
```
[the remark this answers is not in the document]
Comment (qingqing01) [reply doc/design/profiler.md]: Done. Thanks!

File: doc/design/profiler.md (hunk 4 above)
Comment (qingqing01) [reply doc/design/profiler.md]: Done.

File: doc/design/profiler.md (hunk 5 above)
Comment (qingqing01) [reply doc/design/profiler.md]: Done.

File: doc/design/profiler.md (hunk 6 above)
Comment (qingqing01) [reply doc/design/profiler.md]: Done.

File: doc/design/profiler.md (hunk 7 above)
Comment (qingqing01) [reply doc/design/profiler.md]: Done.

File: doc/design/profiler.md (hunk 8 above)
Comment (qingqing01) [reply doc/design/profiler.md]: Done.

File: doc/design/profiler.md (hunk 9 above)
Comment (qingqing01) [reply doc/design/profiler.md]: Done. Thanks!

File: doc/design/profiler.md (hunk 10 above)
Comment (qingqing01) [reply doc/design/profiler.md]: Done.

File: doc/design/profiler.md (hunk 11 above)
Comment (qingqing01) [reply doc/design/profiler.md]: Done. Thanks!

File: doc/design/profiler.md (hunk 2 above)
Comment (qingqing01) [reply doc/design/profiler.md]: Done.

File: doc/design/profiler.md (hunk 1 above)
Comment (qingqing01) [reply doc/design/profiler.md]: Add more comments.

File: doc/design/profiler.md (hunk 12 above)
Comment (qingqing01) [reply doc/design/profiler.md]: Done.

File: doc/design/profiler.md (hunk 13 above)
Comment (qingqing01) [reply doc/design/profiler.md]: Done.

File: doc/design/profiler.md (hunk 14 above)
Comment (qingqing01) [reply doc/design/profiler.md]: Done.

File: doc/design/profiler.md (hunk 3 above)
Comment (qingqing01) [reply doc/design/profiler.md]: Done. And also update in the code.

**Review by kuke — APPROVED** [review]
LGTM

**Pull request merged**